### PR TITLE
Revert "Revert "[cxx-interop] Import custom `NS_OPTIONS` correctly""

### PR DESCRIPTION
**Pull request by egorzhdan**
This re-lands https://github.com/apple/swift/pull/67865 after it got reverted in https://github.com/apple/swift/pull/68435 due to a regression when emitting a pch file.

rdar://112225263

### CODE REVIEW

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Review by zoecarver — APPROVED** [review]
I think @DougGregor should look over this. But this LGTM. Moving `getTypeDefForCXXCFOptionsDefinition` into `ASTContext` is the fix we all discussed.

File: lib/AST/ASTContext.cpp [hunk 1]
```diff
@@ -5732,6 +5732,48 @@ ASTContext::getClangTypeForIRGen(Type ty) {
   return getClangTypeConverter().convert(ty).getTypePtrOrNull();
 }
 
+const clang::TypedefType *ASTContext::getTypedefForCXXCFOptionsDefinition(
```
Comment (DougGregor) [inline lib/AST/ASTContext.cpp:5735]: I'd prefer not to have Clang-specific APIs on `ASTContext`. Can we keep this within the ClangImporter, e.g., by making it a static method there?

**Review by DougGregor — CHANGES_REQUESTED** [review]

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test macOS

**Comment by egorzhdan** [comment]
@swift-ci please smoke test Linux

File: lib/AST/ASTContext.cpp (hunk 1 above)
Comment (egorzhdan) [reply lib/AST/ASTContext.cpp]: Sounds good, I made it a static method of `ClangImporter`

**Review by DougGregor — APPROVED** [review]

**Pull request merged**